src/edge_kernels/VOFAdvectionEdgeAlg.C

```diff
@@ -39,27 +42,51 @@ VOFAdvectionEdgeAlg::VOFAdvectionEdgeAlg(
   massFlowRate_ = get_field_ordinal(
     meta, (useAverages) ? "average_mass_flow_rate" : "mass_flow_rate",
     stk::topology::EDGE_RANK);
-  density_ =
-    get_field_ordinal(realm.meta_data(), "density", stk::mesh::StateNP1);
+  massForcedFlowRate_ =
+    get_field_ordinal(meta, "mass_forced_flow_rate", stk::topology::EDGE_RANK);
+  density_ = get_field_ordinal(meta, "density", stk::mesh::StateNP1);
+  velocity_n_ = get_field_ordinal(meta, "velocity", stk::mesh::StateN);
+
+  std::map<PropertyIdentifier, MaterialPropertyData*>::iterator itf =
+    realm_.materialPropertys_.propertyDataMap_.find(DENSITY_ID);
+
+  if (itf == realm_.materialPropertys_.propertyDataMap_.end()) {
+    density_liquid_ = 1000.0;
+    density_gas_ = 1.0;
+  } else {
+    auto mdata = (*itf).second;
+
+    switch (mdata->type_) {
+    case CONSTANT_MAT: {
+      density_liquid_ = mdata->constValue_;
+      density_gas_ = mdata->constValue_;
+      break;
+    }
+    case VOF_MAT: {
+      density_liquid_ = mdata->primary_;
+      density_gas_ = mdata->secondary_;
+      break;
+    }
+    default: {
+      throw std::runtime_error("Incorrect density property set for VOF "
+                               "calculations. Use a constant or "
+                               "VOF property for density.");
+      break;
+    }
+    }
+  }
 }
 
 void
 VOFAdvectionEdgeAlg::execute()
```

Comment:
It would be good to add some comments to acknowledge the paper where this advection scheme came from and to note and differences during the implementation.  Since we were one deep on the development I'd like to be a little more pedantic about it for this particular case.